pkg/infra/iac2/templates_compiler.go

```diff
@@ -102,43 +110,103 @@ func (tc templatesCompiler) RenderImports(out io.Writer) error {
 	return nil
 }
 
-func (tc templatesCompiler) renderResource(out io.Writer, resource graph.Identifiable) error {
-	// TODO: for now, assume a nice little tree
+func (tc templatesCompiler) renderResource(out io.Writer, resource core.Resource) error {
+
+	tmpl, err := tc.GetTemplate(resource)
+	if err != nil {
+		return err
+	}
+
 	errs := multierr.Error{}
 
+	resourceVal := reflect.ValueOf(resource)
+	for resourceVal.Kind() == reflect.Pointer {
+		resourceVal = resourceVal.Elem()
+	}
 	inputArgs := make(map[string]string)
-	for fieldName, child := range getStructValues(resource) {
-		childType := reflect.TypeOf(child)
-		switch childType.Kind() {
-		case reflect.String:
-			inputArgs[fieldName] = quoteTsString(child.(string))
-		case reflect.Struct, reflect.Pointer:
-			if typedChild, ok := child.(graph.Identifiable); ok {
-				inputArgs[fieldName] = tc.getVarName(typedChild)
-			} else {
-				errs.Append(errors.Errorf(`child struct of %v is not of a known type: %v`, resource, child))
-			}
-		default:
-			errs.Append(errors.Errorf(`unrecognized input type for %v [%s]: %v`, resource, fieldName, child))
+	var zeroValue reflect.Value
+	for fieldName := range tmpl.inputTypes {
+		childVal := resourceVal.FieldByName(fieldName)
+		if childVal == zeroValue {
+			zap.S().Warnf(
+				`Klotho compiler error: no field %s.%s while rendering typescript template`,
+				resourceVal.Type().Name(),
+				fieldName)
+			continue
 		}
+		resolvedValue := tc.resolveStructInput(childVal)
+		if resolvedValue == "" {
+			errs.Append(errors.Errorf(`child struct of %v is not of a known type: %v`, resource, childVal.Interface()))
+		} else {
+			inputArgs[fieldName] = resolvedValue
+		}
+
 	}
 
 	varName := tc.getVarName(resource)
-	tmpl, err := tc.GetTemplate(resource)
-	if err != nil {
-		return err
-	}
 
 	fmt.Fprintf(out, `const %s = `, varName)
 	errs.Append(tmpl.RenderCreate(out, inputArgs))
-	_, err = out.Write([]byte(";\n"))
+	_, err = out.Write([]byte(";"))
 	if err != nil {
 		return err
 	}
 
 	return errs.ErrOrNil()
 }
 
+// resolveStructInput translates a value to a form suitable to inject into the typescript as an input to a function.
+func (tc templatesCompiler) resolveStructInput(childVal reflect.Value) string {
```

Comment:
Isn't this just writing to JSON? Kind of hard to tell, but seems like it. (JSON being the "JavaScript Object Notation" and all, it should be what we want)

Reply:
Very close, but not quite. The two big differences are:

- for structs (and pointers to structs), including ones deep in arrays/maps, we render them as the variable name that the struct had been rendered into (`tc.getVarName(typedChild)`)
- for strings, we use backtick strings instead of json strings, so that multiline messes like policy documents render nicely; otherwise, instead of
  ```
  assumeRolePolicy: `{
      Version: '2012-10-17',
      Statement: [
        {
          Action: 'sts:AssumeRole',
          Principal: {
            Service: 'lambda.amazonaws.com',
          },
          Effect: 'Allow',
          Sid: '',
        },
      ],
    }`,
  ```
  ... you get:
  ```
  assumeRolePolicy: "{\n\tVersion: '2012-10-17',\n\tStatement: [\n\t\t{\n\t\t\tAction: 'sts:AssumeRole',\n\t\t\tPrincipal: {\n\t\t\t\tService: 'lambda.amazonaws.com',\n\t\t\t},\n\t\t\tEffect:
'Allow',\n\t\t\tSid: '',\n\t\t},\n\t],\n}"
  ```

Reply:
Makes sense 👍 (could that description go in the godocs)

For the multiline stuff, we'll likely need an indentation-level for true pretty printing, but good enough for now definitely.

Reply:
Will definitely add it!